cpp/src/arrow/dataset/dataset.h

```diff
@@ -191,9 +200,9 @@ class ARROW_DS_EXPORT InMemoryDataset : public Dataset {
       std::shared_ptr<Schema> schema) const override;
 
  protected:
-  Result<FragmentIterator> GetFragmentsImpl(Expression predicate) override;
+  Future<FragmentVector> GetFragmentsImpl(Expression predicate) override;
```

Comment:
I was going to let this be addressed by ARROW-8163.

However, if we're getting ARROW-10882 done first then rather than go with an InMemoryFragment approach I'd recommend switching to AsyncGenerator<Fragment> at that point since we're going to do so sooner or later anyways.

Reply:
Also, it hopefully won't be too much work.  The scanner already consumes the FragmentVector as a generator so the only change needed would be to push the MakeVectorGenerator down into the file formats.